src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp

```diff
@@ -1331,6 +1324,118 @@ void ov::npuw::LLMCompiledModel::gemma_transformations(const std::shared_ptr<ov:
+    const uint32_t min_response_len = align_to(m_cfg.get<::intel_npu::NPUW_LLM_MIN_RESPONSE_LEN>(), 64u);
+    const uint32_t max_generation_token_len = m_kvcache_desc.max_generation_token_len;
+    const bool enable_kvcache_variants = m_cfg.get<::intel_npu::NPUW_LLM_KVCACHE_VARIANTS>();
+
+    // Check if KV cache variants feature is enabled
+    if (enable_kvcache_variants) {
+        LOG_INFO("KV cache variants feature is ENABLED");
+        LOG_INFO(
+            "Creating multiple KV cache model variants with stepping: 1K+min_response_len, 2K+min_response_len, etc.");
+
+        // Determine KV cache size steps: (1K + min_response_len), (2K + min_response_len), (4K + min_response_len), (8K
+        // + min_response_len), etc.
+        std::vector<uint32_t> kv_size_steps;
+        for (uint32_t base_size = 1024; base_size + min_response_len <= total_kv_size; base_size *= 2) {
+            kv_size_steps.push_back(base_size + min_response_len);
+        }
+        // Always include the total size if it's not already in the list
+        if (kv_size_steps.empty() || kv_size_steps.back() < total_kv_size) {
+            kv_size_steps.push_back(total_kv_size);
+        }
+
+        LOG_DEBUG("KV cache size variants: ");
+        for (const auto& size : kv_size_steps) {
+            LOG_DEBUG("  - " << size);
+        }
+
+        // Store the sizes for runtime selection
+        m_kvcache_sizes = kv_size_steps;
+    } else {
+        LOG_INFO("KV cache variants feature is DISABLED - using single model");
+        // Use only the total size (traditional single-model approach)
+        m_kvcache_sizes = {total_kv_size};
+    }
+
+    // Generate KV cache model variants
+    LOG_INFO("Generating " << m_kvcache_sizes.size() << " KV cache model variants...");
+    std::vector<std::shared_ptr<ov::Model>> kvcache_variant_models;
+    kvcache_variant_models.reserve(m_kvcache_sizes.size());
+
+    for (size_t i = 0; i < m_kvcache_sizes.size(); ++i) {
+        const uint32_t kv_size = m_kvcache_sizes[i];
+
+        if (kv_size == total_kv_size) {
+            // Last variant uses the main model directly - reshape it first
+            LOG_DEBUG("Variant " << (i + 1) << "/" << m_kvcache_sizes.size() << " (size=" << kv_size
+                                 << "): using and reshaping main model");
+            reshape_to_static(kvcache_model,
+                              max_generation_token_len,
+                              total_kv_size,
+                              axes,
+                              m_max_lora_rank,
+                              whisper_lhs_seq_size);
+            kvcache_variant_models.push_back(kvcache_model);
+        } else {
+            // Clone and create smaller variants
+            LOG_DEBUG("Variant " << (i + 1) << "/" << m_kvcache_sizes.size() << " (size=" << kv_size
+                                 << "): cloning and reshaping");
+            auto kvcache_variant = kvcache_model->clone();
```

Comment:
would each model get its unique name? would it make sense to add a certain size to the model name? As the same generation config is passed to each variant, their subgraphs may be dumped all at once.